e2e/issues_test.go

```diff
@@ -0,0 +1,81 @@
+// +build e2e
+
+//
+
+package e2e
+
+import (
+	"bytes"
+	"context"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/hashicorp/consul-terraform-sync/config"
+	"github.com/hashicorp/consul-terraform-sync/controller"
+	"github.com/hashicorp/consul-terraform-sync/event"
+	"github.com/hashicorp/consul-terraform-sync/testutils"
+	"github.com/stretchr/testify/require"
+)
+
+// issues GH-236 and GH-239
+func TestServicesRenderRace(t *testing.T) {
+	const NumberOfServices = 100
+
+	srv := testutils.NewTestConsulServer(t,
+		testutils.TestConsulServerConfig{HTTPSRelPath: "../testutils"})
+	defer srv.Stop()
+
+	tempDir := t.Name()
+	cleanup := testutils.MakeTempDir(t, tempDir)
+	//_ = cleanup
+	defer cleanup()
+
+	ctx, ctxCancel := context.WithCancel(context.Background())
+	defer ctxCancel()
+
+	// seed server with test data
+	testServices := testutils.TestServices(NumberOfServices)
+	testutils.AddServices(t, srv, testServices)
+
+	// config
+	conf := config.DefaultConfig()
+	conf.BufferPeriod.Enabled = config.Bool(false)
+	serviceNames := make([]string, NumberOfServices)
+	for i, s := range testServices {
+		serviceNames[i] = s.Name
+	}
+	conf.Tasks = &config.TaskConfigs{
+		&config.TaskConfig{
+			Name:     config.String("serv_rend_race_task"),
+			Source:   config.String("../../test_modules/null_resource"),
+			Services: serviceNames,
+		}}
+	conf.Consul.Address = config.String(srv.HTTPAddr)
+	conf.Finalize()
+	conf.Driver.Terraform.WorkingDir = config.String(tempDir)
+	path, err := filepath.Abs("./")
+	require.NoError(t, err)
+	conf.Driver.Terraform.Path = config.String(path)
+
+	// download terraform
+	err = controller.InstallDriver(ctx, conf)
+	require.NoError(t, err)
+
+	// run controller
+	ctrl, err := controller.NewReadWrite(conf, event.NewStore())
+	rwCtrl := ctrl.(*controller.ReadWrite)
+	require.NoError(t, err)
+	_, err = rwCtrl.Init(ctx)
+	require.NoError(t, err)
+	err = rwCtrl.Once(ctx)
+	require.NoError(t, err)
```

Comment:
I'm curious about choosing to run CTS by calling the controller rather than running the binary! I know we have the runSyncOnce() option which could be useful. I was wondering if you could share your thought process?

Reply:
Couple reasons for not calling cli.Run() directly like runSyncOnce()..
- This is a more targeted test of the one problem. I wanted to constrain it more to the problem area.
- I wanted to try it. Kim's been doing this and it seemed nice and explicit as you don't use intermediate config files, but create the config object directly.

Why I didn't run the binary.. I was thinking of proposing we switch from the binary over to using cli.Run() instead. It would mean no more worrying about rebuilding the binary each time you make some change and are running individual tests. It would also eliminate the binary call overhead which should shave a bit of time off the suite for no real loss as all that really tests is the underlying system exec call and Go's os.Args parsing.

Reply:
Thanks for sharing! That makes sense. I also prefer how config building works for cli.Run(). I'm up for switching tests that target constrained problems/use-cases to using cli.Run()! We can use running the full binary for specific tests that actually benefit from it and maybe some 'basic' tests (for peace of mind for a complete e2e).